Overhaul comment attachment

JLHwung opened the pull request:
<!--
Before making a PR, please read our contributing guidelines
https://github.com/babel/babel/blob/main/CONTRIBUTING.md

Please note that the Babel Team requires two approvals before merging most PRs.

For issue references: Add a comma-separated list of a [closing word](https://help.github.com/articles/closing-issues-via-commit-messages/) followed by the ticket number fixed by the PR. (it should be underlined in the preview if done correctly)

If you are making a change that should have a docs update: submit another PR to https://github.com/babel/website
-->

| Q                        | A <!--(Can use an emoji 👍) -->
| ------------------------ | ---
| Fixed Issues?            | Some comments after `}` / `+` are mis attached as leading/trailing comments. Fixes #11576, closes #12560
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | <!-- If only readme change, add `[skip ci]` to your commits -->
| Any Dependency Changes?  |
| License                  | MIT

<!-- Describe your changes below in as much detail as possible -->

## Abstract

This PR overhauls current comment attachments. The updated test fixtures are all bugfixes. It is faster than current approach in general cases.

## [Design docs](https://github.com/JLHwung/babel/blob/overhaul-comment-attachment/packages/babel-parser/ast/comment-attachment.md)

## Benchmark Results
<details>
  <summary>leading comments + trailing comments (common use case, O(n^2) -> O(n))</summary>

```js
// c
a
// c
a
// c
...
a
```

```
baseline 128 leading comments + 127 trailing comments: 4_155 ops/sec ±2.05% (0.241ms)
baseline 256 leading comments + 255 trailing comments: 1_792 ops/sec ±2.07% (0.558ms)
baseline 512 leading comments + 511 trailing comments: 669 ops/sec ±2.2% (1.495ms)
baseline 1024 leading comments + 1023 trailing comments: 217 ops/sec ±2.08% (4.599ms)
current 128 leading comments + 127 trailing comments: 9_686 ops/sec ±56.64% (0.103ms)
current 256 leading comments + 255 trailing comments: 6_698 ops/sec ±0.52% (0.149ms)
current 512 leading comments + 511 trailing comments: 3_303 ops/sec ±0.38% (0.303ms)
current 1024 leading comments + 1023 trailing comments: 1_535 ops/sec ±1.84% (0.651ms)
```
</details>

<details>
  <summary>leading comments (10%)</summary>

```js
// c
// c
...
// c
{}
```

```
baseline 128 leading comments: 77_947 ops/sec ±6.78% (0.013ms)
baseline 256 leading comments: 43_160 ops/sec ±2.44% (0.023ms)
baseline 512 leading comments: 22_572 ops/sec ±2.22% (0.044ms)
baseline 1024 leading comments: 10_918 ops/sec ±2.47% (0.092ms)
current 128 leading comments: 83_604 ops/sec ±8.04% (0.012ms)
current 256 leading comments: 46_832 ops/sec ±2.47% (0.021ms)
current 512 leading comments: 26_373 ops/sec ±1.53% (0.038ms)
current 1024 leading comments: 12_952 ops/sec ±2.26% (0.077ms)
```
</details>

<details>
  <summary>nested leading comments (O(n^2) -> O(n))</summary>

```js
// c
{
// c
{
...
// c
{}} ... }
```

```
baseline 128 nested leading comments: 9_454 ops/sec ±35.7% (0.106ms)
baseline 256 nested leading comments: 4_724 ops/sec ±1.6% (0.212ms)
baseline 512 nested leading comments: 1_755 ops/sec ±2.16% (0.57ms)
baseline 1024 nested leading comments: 348 ops/sec ±2.33% (2.871ms)
current 128 nested leading comments: 11_226 ops/sec ±32.57% (0.089ms)
current 256 nested leading comments: 6_470 ops/sec ±1.47% (0.155ms)
current 512 nested leading comments: 2_834 ops/sec ±0.63% (0.353ms)
current 1024 nested leading comments: 1_380 ops/sec ±0.99% (0.725ms)
```
</details>

<details>
  <summary>trailing comments (marginal improvements)</summary>

```js
{{ ... {}// c
} // c
} // c
...
} // c
```

```
baseline 128 trailing comments: 83_635 ops/sec ±8.24% (0.012ms)
baseline 256 trailing comments: 40_747 ops/sec ±2.67% (0.025ms)
baseline 512 trailing comments: 24_274 ops/sec ±2.11% (0.041ms)
baseline 1024 trailing comments: 10_469 ops/sec ±3.81% (0.096ms)
current 128 trailing comments: 73_240 ops/sec ±11.22% (0.014ms)
current 256 trailing comments: 42_596 ops/sec ±3.48% (0.023ms)
current 512 trailing comments: 25_424 ops/sec ±2.47% (0.039ms)
current 1024 trailing comments: 13_662 ops/sec ±0.98% (0.073ms)
```
</details>

<details>
  <summary>nested trailing comments (25% slower) </summary>

```js
{{ ... {
} // c
} // c
...
} // c
```

```
baseline 128 nested trailing comments: 11_380 ops/sec ±52.74% (0.088ms)
baseline 256 nested trailing comments: 8_513 ops/sec ±0.63% (0.117ms)
baseline 512 nested trailing comments: 3_777 ops/sec ±1.83% (0.265ms)
baseline 1024 nested trailing comments: 1_806 ops/sec ±0.97% (0.554ms)
current 128 nested trailing comments: 10_530 ops/sec ±39.15% (0.095ms)
current 256 nested trailing comments: 6_344 ops/sec ±1.14% (0.158ms)
current 512 nested trailing comments: 3_064 ops/sec ±1.05% (0.326ms)
current 1024 nested trailing comments: 1_332 ops/sec ±1.03% (0.751ms)
```

I think it is acceptable since trailing-only comments is rare. Most comments are both leading / trailing comments of adjacent AST nodes.
</details>

<details>
  <summary>inner comments (10%) </summary>

```js
[ // c
// c
...
// c
]
```

```
baseline 128 inner comments: 74_560 ops/sec ±9.44% (0.013ms)
baseline 256 inner comments: 43_778 ops/sec ±1.93% (0.023ms)
baseline 512 inner comments: 23_776 ops/sec ±1.6% (0.042ms)
baseline 1024 inner comments: 11_676 ops/sec ±1.5% (0.086ms)
current 128 inner comments: 86_420 ops/sec ±9.67% (0.012ms)
current 256 inner comments: 48_437 ops/sec ±1.56% (0.021ms)
current 512 inner comments: 25_333 ops/sec ±1.42% (0.039ms)
current 1024 inner comments: 13_312 ops/sec ±2.39% (0.075ms)
```
</details>

<details>
  <summary>nested inner comments (current main incorrectly produces leading comments) </summary>

```js
[, // c
[, // c
...
[, // c
]] ... ]
```

```
baseline 128 nested inner comments: 6_698 ops/sec ±66.03% (0.149ms)
baseline 256 nested inner comments: 4_459 ops/sec ±1.21% (0.224ms)
baseline 512 nested inner comments: 1_565 ops/sec ±1.62% (0.639ms)
baseline 1024 nested inner comments: 519 ops/sec ±2% (1.927ms)
current 128 nested inner comments: 8_711 ops/sec ±63.38% (0.115ms)
current 256 nested inner comments: 6_451 ops/sec ±1.39% (0.155ms)
current 512 nested inner comments: 3_061 ops/sec ±0.85% (0.327ms)
current 1024 nested inner comments: 1_276 ops/sec ±1.18% (0.784ms)
```
</details>

<details>
  <summary>many identifiers (without comments, 15%)</summary>

```js
a;a; ... a;
```

```
baseline 64 length-1 identifiers: 18_321 ops/sec ±84.1% (0.055ms)
baseline 128 length-1 identifiers: 17_076 ops/sec ±1.72% (0.059ms)
baseline 256 length-1 identifiers: 8_546 ops/sec ±1.64% (0.117ms)
baseline 512 length-1 identifiers: 4_339 ops/sec ±2.54% (0.23ms)
baseline 1024 length-1 identifiers: 2_185 ops/sec ±1.39% (0.458ms)
current 64 length-1 identifiers: 23_844 ops/sec ±82.38% (0.042ms)
current 128 length-1 identifiers: 21_509 ops/sec ±1.75% (0.046ms)
current 256 length-1 identifiers: 9_722 ops/sec ±3.6% (0.103ms)
current 512 length-1 identifiers: 4_970 ops/sec ±1.77% (0.201ms)
current 1024 length-1 identifiers: 2_404 ops/sec ±1.6% (0.416ms)
```

</details>
<a href="https://gitpod.io/#https://github.com/babel/babel/pull/13521"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

kaicataldo reviewed (approved):
Amazing work! 🎉

JLHwung commented on packages/babel-parser/test/fixtures/comments/basic/call-expression-no-argument/input.js:
```diff
@@ -0,0 +1 @@
+foo /* 1 */ (/* 2 */)
```
The comment `/* 2 */` is now an `innerComment` of `ExpressionStatement`, which is not bad since 1) it is lost on current main and 2) we don't have AST structures for the bracket token.

JLHwung commented on packages/babel-parser/src/parser/comments.js:
```diff
@@ -1,288 +1,214 @@
+ * @property {Array<Comment>} comments - the containing comments
+ * @property {Node | null} leadingNode - the immediately preceding AST node of the whitespace token
+ * @property {Node | null} trailingNode - the immediately following AST node of the whitespace token
+ * @property {Node | null} containerNode - the outermost AST node containing the whitespace
```
I think in the future we can change the behaviour to have `innerComments` attached to the _innermost_ AST node. So in the example `foo (/* 2 */)`, `/* 2 */` will the an `innerComments` of call expression. The idea is to provide the generator more information about how to insert `innerComments`.

Or should we change in this PR now?

kaicataldo commented:
Sorry, didn't realize it wasn't ready for review 😅

JLHwung commented:
Converted back to draft as I just find new bugs: e.g. Babel does not attach comments after a DirectiveLabel:

```js
"use strict"/* foo */;
```

KFlash commented:
@JLHwung Kataw have loads of comment tests you can find [here](https://github.com/kataw/kataw/tree/main/test/__snapshot__/ecma262/printer/comments). In fact Kataw is the only one that get all comments 100%, so I still think you will have issues with class semicolon and elisons in array literal and pattern.  There are also some other edge cases.

KFlash commented:
@JLHwung I forgot that you may suffer with trailing comments if Babel in the future should allow optional trailing comments as in Prettier. 
You may need to extend your algorithm to fix this. `[a/*1*/]` and `[a,/*1*/]`

Typescript have a few edge cases with and without comments where trailing comma is required. Not sure how Babel works around it.

nicolo-ribaudo reviewed (commented):
I only read the PR description so far. We should probably link this PR in the code, or add a `src/parser/comments.md` file, so that whenever someone will have to modify the algorithm they can read the rationale behind it.

1. (P2) should probably be
   ```
   w1.start ≤ w2.start ≤ w1.end
   ```

2. What does (P3) mean for an input code which is just `foo` (without any spaces or newlines)?

nicolo-ribaudo commented on packages/babel-parser/src/parser/comments.js:
```diff
@@ -284,4 +290,104 @@ export default class CommentsParser extends BaseParser {
+            unattachedCommentStack.splice(i, 1);
+          } else if (node.type !== "Program") {
+            // we have a node share the same length of containerNode, but its finishNode is invoked later
+            // than containerNode, so this node is the outer node. E.g. ExpressionStatement contains a VariableDeclaration
```
Nit: `ExpressionStatement` cannot contain a `VariableDeclaration`, but maybe an `AssignmentExpression`?

JLHwung replied:
Yes! You are right. This branch is removed so we are all good here.

nicolo-ribaudo commented on packages/babel-parser/src/parser/comments.js:
```diff
@@ -284,4 +290,104 @@ export default class CommentsParser extends BaseParser {
+            case "ObjectPattern":
+              this.adjustInnerComments(node, node.properties, comments);
+              break;
+            case "CallExpression":
```
We also need to do this for `OptionalCallExpression`, and maybe for function definitions?

JLHwung replied:
Yes it is addressed in https://github.com/babel/babel/pull/13521/commits/e5f09a5a82cb3b31906a4168a4eff16815fec604.

Just in case, are you reviewing on old diff?

nicolo-ribaudo replied:
Uh I think I was accidentally reviewing a specific commit :sweat_smile:

KFlash commented:
The main issue is the performance. I see the benchmarks, but in real life. How many source codes per second does this generator perform? And how much memory is consumed?

Seems to be too much. Still using a class to handle comments, unnecessary that could have been done in one go etc.

KFlash commented:
Playing around with Babel REPL and found lots of comment issues

Here is a few

```ts
switch (c) { /*1*/ } /*2*/
switch (c) { /*1*/ } /*2*/
switch (c)/*1*/ { /*2*/ } /*3*/
x(/*1*/)
(a(/*1*/))
(a())(a(/*1*/))
[,,,/*1*/,,,,]

// should not break 
"string"; /*1*/
```

JLHwung commented:
@nicolo-ribaudo 

> What does (P3) mean for an input code which is just foo (without any spaces or newlines)?

Good question! In that case the list of comment whitespaces is an empty set. P3 is then deduced to nothingness.

> We should probably link this PR in the code, or add a src/parser/comments.md file

I can add it to the `docs/` directory.

@KFlash 

> How many source code files per second does this generator perform?

I run the performance test on https://github.com/babel/parser_performance. When parsing `es5/angular`, this branch is faster than 7.14.7. The repo has not been updated for long, I should have used `seafox` here.

| fixture | acorn | Babel 7.14.7 | This branch | meriyah
| --- | --- | --- | --- | --- |
| es5/angular.js  | 28.51 ops/sec (27ms) | 17.41 ops/sec (38ms) | 24.19 ops/sec (28ms) | 28.21 ops/sec (24ms)
 
I have added a new identifier benchmark result (without any comments). Because the overhead of comment tracking is reduced, the new algorithm is 15% faster than current one. The identifier benchmark predict the performance improvements on `es5/angular`.

> Playing around with Babel REPL and found lots of comment issues

Thanks! I checked the cases and the comments are now all attached to AST but `@babel/generator` does not print them. Before this PR the `innerComments` does not work quite well so I won't be surprised if `@babel/generator` does not print them. The issue will be addressed in separate PRs.

<blockquote>
<pre><code>// should not break 
"string"; /*1*/
</code></pre>
</blockquote>

Why? I think they are equivalent. Babel generator does not preserve the whitespace.

KFlash commented:
@JLHwung This case is a trailing of the empty stmt. When removed it should still be the same, but be a "trailing of string literal".  In current REPL  a line break is inserted and the comment is no longer a trailing comment. It's a detached comment.

```ts
// should not break 
"string"; /*1*/
```
Why using Acorn and Meriyah in the benchmark? They doesn't have any printer / generator support. And this PR is about Babel generator and internal code?

Why not add Kataw to the benchmark? The printer is located [here](https://kataw.github.io/kataw/kataw_print/).

JLHwung commented:
@KFlash This PR is focused on parser because before this PR we failed to attach comments in edge cases, so we should prioritize the parser part. The generator support would be addressed later.

Babel parses `;` as empty statement. So 
```
; /*1*/
```
is an empty statement with trailing comments.

KFlash commented:
The Kataw parser can be located [here](https://kataw.github.io/kataw/kataw_ast/) in case you want to add it ;)

Regarding the comment. It was printed wrong in the REPL but maybe parsed correctly. Compare against Prettier. They also handle it as an trailing comment. Not a detached one.

JLHwung commented:
@KFlash I think the comment attachment can only be perfectly handled on a CST where every non-whitespace token has a node representation to which a comment can be attached. The ideal implementation should be language-agnostic as long as space and non-space are well defined. 

If Babel had sort of `TrailingCommaWrapper` AST node, we could get rid of the current adjustments for inner comments. Here it is the extra cost of addressing comment attachment on AST.

KFlash commented:
After my opinion there is a "design flaw" in all AST parser when it comes to loc tracking. It shouldn't count whitespace at all. That way you have more control, and you can use the start / end value of each AST / CST node in the printer to use a separated WS skipping that starts at given position and ends if hit a token that is not whitespace. 
Then there is no need to attach this to AST either and you can very easy collect all comments that is between two tokens except for "lists". Lists is defined in the ECMA specs and you would need an extra AST node to get correct loc position after the list token has been consumed. E.g. '(', '{' etc.

This is out of scope for Babel, but the only way you can get comments 100% correct.

KFlash commented:
@JLHwung Here is close to [200 comment tests](https://github.com/kataw/kataw/tree/main/test/__snapshot__/ecma262/printer/comments). This PR should parse them all.

@nicolo-ribaudo Did you validate all possible comment attachment combinations before approved this PR? Check the test I linked too. This PR still fails on 2 out of 5 of this tests, so there are still things to fix.

JLHwung commented:
@KFlash Can you offer a list of failing tests? If the comment is _not_ attached to AST nodes, it should be addressed in this PR. Otherwise it is a generator bug and will be addressed later.

KFlash commented:
@JLHwung Just test all the 200 tests I linked too. You can see if the comments are attached or not. As said earlier 2 out of 5 cases the comments are not attached.
Why do you need an list? Just add all the tests to Babel and you can see witch one is failing.

JLHwung commented:
> You can find a sub-folder in the tests I linked too - "Babel issues". I haven't tested if they are fixed with this PR

Thanks. I have checked all the babel issues and they are already fixed. The only exception is #11971: In this case the parser does attach comments, but traverser fails to preserve comments when the nodes are manipulated.

> Just test all the 200 tests I linked too. You can see if the comments are attached or not.

Inspired from your test cases, I do find an issue not yet fixed in this PR: It is also failing on the main branch.

```js
(class { async /* 1 */ *f () {} })
```

The `/* 1 */` comment should be the `innerComments` of the class method. However, Babel fails to attach because we parse `async` as an identifier and then reinterpret as an async generator method. Because trailing comments are prioritized over innerComments, in this case the comment is attached to the `async` identifier node, which apparently does not exist in the final AST.

JLHwung commented on packages/babel-parser/src/parser/comments.js:
```diff
@@ -210,4 +210,35 @@ export default class CommentsParser extends BaseParser {
+   * @returns
+   * @memberof CommentsParser
+   */
+  resetPreviousNodeTrailingComments(node: N.Node) {
```
I don't like this routine. It is a workaround for the fact that we are using `parseIdentifier` as a lookahead.

nicolo-ribaudo commented:
> Good question! In that case the list of comment whitespaces is an empty set. P3 is then deduced to nothingness.

However, the list of nodes is not empty: you have at least a node for which P3 is false exactly because the list of whitespace comments is empty (i.e. at least a node to which you cannot associate a whitespace comment).

JLHwung commented:
> The list of nodes is not empty

Oh, P3 does not account for non-whitespace AST nodes. It says for every _comment_ node `c`, we have whitespace `w` covering `c`. So it is sufficient to handle comment whitespaces. If there is no comment in a source, P3 deduces to nothingness.

nicolo-ribaudo commented:
Ohhh I misread it.

JLHwung commented on packages/babel-parser/test/fixtures/comments/basic/async-arrow-function/input.js:
```diff
@@ -0,0 +1,3 @@
+/* 1 */ async /* 2 */ (/* 3 */) /* 4 */=>x;
```
The current main fails to attach `/* 2 */` and `/* 3 */`, and incorrectly attaches `/* 4 */` to leading comments of `x`.

JLHwung commented on packages/babel-parser/test/fixtures/comments/basic/class-static-block/input.js:
```diff
@@ -0,0 +1 @@
+(class /* 1 */ { /* 2 */ static /* 3 */ {/* 4 */} /* 5 */ } /* 6 */);
```
Currently Babel does not attach comments `3`, `4` and `5`. It is now fixed in this PR.

JLHwung commented on packages/babel-parser/test/fixtures/comments/basic/class-method-static-generator/input.js:
```diff
@@ -0,0 +1,2 @@
+(class /* 1 */ { /* 2 */ static /* 3 */ * /* 4 */ f /* 5 */ (/* 6 */) /* 7 */ { /* 8 */ } /* 9 */ } /* 10 */);
```
Currently Babel attaches comments `1` to `7` to the method's block statement (horribly). It is now fixed in this PR.

KFlash commented:
@JLHwung Did you 'adopt' some of Kataw's comment tests?  Some of the tests looks familiar. 
How are you handling class semicolon comments?

JLHwung commented:
@KFlash

> Did you 'adopt' some of Kataw's comment tests?

Yes. Tests added in https://github.com/babel/babel/pull/13521/commits/2c6bc30682fcb156bffcc5427ea6bf46e3a82662 are derived from Kataw tests. 

> How are you handling class semicolon comments?

They are all `innerComments` of the `ClassBody` since Babel does not create an AST for semicolons here.

KFlash commented:
In the Kataw repo there also exist [comments for types](https://github.com/kataw/kataw/tree/main/test/__snapshot__/types/printer/comments). Did you look into them?  And more tests will be added soon because I'm sure there are edge cases.

JLHwung commented on packages/babel-parser/test/fixtures/comments/basic/async-do-expression/output.json:
```diff
@@ -0,0 +1,90 @@
+        "expression": {
+          "type": "DoExpression",
+          "start":19,"end":37,"loc":{"start":{"line":1,"column":19},"end":{"line":1,"column":37}},
+          "leadingComments": [
```
This is an unrelated bug: An async do expression should start at the position of `async`, here it starts at `do`. I will fix this issue in separate PR and rebase this one on top of that.

JLHwung commented on packages/babel-parser/test/fixtures/comments/basic/async-function/input.js:
```diff
@@ -0,0 +1 @@
+/* 1 */async/* 2 */function f () {}
```
Currently Babel attaches `/* 2 */` to `f`, it is fixed in this PR.

kaicataldo commented on packages/babel-parser/ast/comment-attachment.md:
```diff
@@ -0,0 +1,129 @@
+# Comment attachment
```
❤️
